### PR TITLE
Datepicker with 12 hour time format midnight error

### PR DESCRIPTION
**Pull request by brianteeman**
PR for #17938

Thanks to @gclw for the code and his patience

### Test instructions
in \administrator\components\com_content\models\forms\article.xml
add the following two lines to the "created" field

`timeformat="12"
format="%Y-%m-%d %H:%M:%S"
`
As you set the time in the calendar observe the time in the field. 
1. Set the time to 4pm and the field shows 16:00
2. Now set the time to 12am and the field will show 12:00

This is the bug. It should show 00:00

Apply this PR,clear cache, and it will behave correctly as in the video below

### after PR

clock

### CODE REVIEW

**Comment by ggppdk** [comment]
if i popup calendar in the hour, minute field if i select

10:00 am it shows 10:00
11:00 am it shows 11:00
12:00 am it shows 00:00

10:00 pm it shows 22:00
11:00 pm it shows 23:00
12:00 pm it shows 12:00

i do not say that this change is wrong
it seems correct

but there is an ux problem, that the hours selector 12 is placed after 11,
thus user selecting would expected it to show something bigger right ?

@dgt41 
maybe replace 12 with 00 when calendar is shown in am / pm format , and of course place 00 on top of the hours drop down selector ?

**Comment by brianteeman** [comment]
That is beyond the scope of this pr

File: media/system/js/fields/calendar.js [hunk 1]
```diff
@@ -178,7 +178,7 @@
 		if (!self.params.time24) {
 			if (/pm/i.test(ampm.value) && hours < 12) {
 				hours = parseInt(hours) + 12;
-			} else if (/am/i.test(ampm.value) && hours === 12) {
+			} else if (/am/i.test(ampm.value) && hours == 12) {
```
Comment (dgrammatiko) [inline media/system/js/fields/calendar.js:181]: @brianteeman please revert the strict check

**Review by dgrammatiko — CHANGES_REQUESTED** [review]

File: media/system/js/fields/calendar.js (hunk 1 above)
Comment (brianteeman) [reply media/system/js/fields/calendar.js:181]: If I do that then it doesnt work :(

**Comment by GCLW** [comment]
@ggppdk 
I think it would be wise to keep the hour selector containing only 1-12 in when using the timeformat=12 attribute. Only when the timeformat=24 is used should there be a "00" in the selector. There is no UX problem. it is a logical range 01-12 that stays consistent, the user only selects if it is AM or PM.

I think you are trying to force the logic that if someone were to select 11 PM and changed their mind to make it midnight they would just bump up the 11 to a 12, which effectively now makes it 12 PM (noon). That is not a UX problem. That is 100% user error.

The correct logic by a user would be "I have it set to 11:30pm, but I need to change it to midnight. The user would then change the 11 to a 12 and change PM to AM. That is the user logic.

**Comment by brianteeman** [comment]
it is out of scope for this Pr anyway

File: media/system/js/fields/calendar.js (hunk 1 above)
Comment (GCLW) [reply media/system/js/fields/calendar.js:181]: It you want to keep the strict compare "===" you would have to use parseInt(). I think the hours var is being recognized as a string.

`} else if (/am/i.test(ampm.value) && parseInt(hours) === 12) {`

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on fae7637162109828f2d8ff897753cdedfc9515e7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18240">issues.joomla.org/tracker/joomla-cms/18240</a>.</sub>

**Comment by ghost** [comment]
marked as "Easy test".<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18240">issues.joomla.org/tracker/joomla-cms/18240</a>.</sub>

**Comment by fancyFranci** [comment]
I have tested this item :white_check_mark: successfully on fae7637162109828f2d8ff897753cdedfc9515e7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18240">issues.joomla.org/tracker/joomla-cms/18240</a>.</sub>

**Comment by ghost** [comment]
RTC after two successful tests.

**Pull request merged**